format-common/src/main/java/io/cdap/plugin/format/charset/fixedlength/FixedLengthCharsetTransformingDecompressorStream.java

```diff
@@ -0,0 +1,119 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.charset.fixedlength;
+
+import io.cdap.plugin.format.charset.CharsetTransformingLineRecordReader;
+import org.apache.hadoop.io.compress.DecompressorStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * DecompressorStream implementation for the FixedLengthCharsetTransformingDecompressor.
+ * <p>
+ * This can be used to convert input streams containing bytes for fixed length charsets into UTF-8 bytes.
+ */
+public class FixedLengthCharsetTransformingDecompressorStream extends DecompressorStream {
+  private static final Logger LOG = LoggerFactory.getLogger(FixedLengthCharsetTransformingDecompressorStream.class);
+
+  //Starting and ending position in the file.
+  protected final long start;
+  protected final long end;
+  protected final FixedLengthCharset fixedLengthCharset;
+
+  protected FixedLengthCharsetTransformingDecompressorStream(InputStream in,
+                                                             FixedLengthCharset fixedLengthCharset,
+                                                             long start,
+                                                             long end)
+    throws IOException {
+    super(in, new FixedLengthCharsetTransformingDecompressor(fixedLengthCharset));
+    long skippedBytes = in.skip(start);
+    this.fixedLengthCharset = fixedLengthCharset;
+    this.start = start;
+    this.end = end;
+  }
+
+  @Override
+  protected int decompress(byte[] b, int off, int len) throws IOException {
+    //Set input for decompression if it's needed for execution.
+    if (this.decompressor.needsInput()) {
+      int l = getCompressedData();
+      if (l > 0) {
+        this.decompressor.setInput(buffer, 0, l);
+      }
+    }
+
+    //Proceed with super method.
+    return super.decompress(b, off, len);
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    // Since we're working with a Charset Transforming decompressor, we can calculate the current position on the
+    // input file.
+    // By adding the starting position in the file with the number of bytes we have read so far.
+    FixedLengthCharsetTransformingDecompressor flcDecompressor =
+      (FixedLengthCharsetTransformingDecompressor) this.decompressor;
+
+    //Actual position is starting possition + the number of bytes we have consumed.
+    return start + flcDecompressor.getNumConsumedBytes();
+  }
+
+  /**
```

Comment:
@chtyim @albertshau @tivv I want to bring your attention to this line function:

The problem I'm facing at this time is that, in order to ensure that we respect the partition boundaries, and process each record exactly once, I need to make sure that I only read the last line that begins in a partition. This means that if a line is "split" by a partition, the worker that was reading the previous partition is the one that reads this line.

The way Hadoop does this for normal text files is to check against the file position in order to make sure the last line has not been read. However, this poses a problem for me, because the compressor reads part of the file in chunks, and the file position does not necessarily align with the position of the last line that was read and decoded.

The block of code I'm describing can be found here: https://github.com/cdapio/hydrator-plugins/blob/83427f7002b016044f5f4606d998b92511c92ceb/format-common/src/main/java/io/cdap/plugin/format/charset/CharsetTransformingLineRecordReader.java#L166

This poses a challenge for me. I see 2 ways to fix this:

1) As you see hire, once I approach the partition boundary, I make sure to only read one character at a time, until the function that reads a line is able to complete successfully, and break out of the loop. I dislike this solution because this means that, as I approach the partition boundary, the performance slows down significantly. 

2) The other solution I saw is to calculate the file position based on reading character-by-character from the input in the decompressor. The problem I see with that is that, for every character I read, I would need to create a String instance, convert to bytes, and fill the output buffer in this way. This would give me the ability to accurately determine my position in the underlying file, but doing the Bytes -> 1 Character long String -> Bytes in UTF-8 charset conversion for every character in my input file is definitely not a performant solution.

Let me know if you can think of a better way to approach this partition boundary issue.

Reply:
Did you try switching to ByteBuffer/CharBuffer? I believe those should remove most of the overhead

Reply:
Also, can't you just scan without decoding until you see a line delimiter byte sequence of EOF ? This way you would know exact length of the last line and decode it fast.

Reply:
The latter may have memory problems if lines are very long, so using buffers instead of strings can be a better solution.

Reply:
The way I see it is, you read into `ByteBuffer`, using a `CharsetDecoder` to decode the bytes into the source charset, and then encode it to UTF-8 `ByteBuffer`. Then you can feed the UTF-8 bytes to the regular Hadoop reader.

Reply:
Thanks @tivv. I was still tying up some things. I'll take a look at both of your suggestions and commit my changes 👀

Reply:
@tivv @chtyim I've reimplemented the logic using ByteBuffer and CharBuffer and it works significantly faster than before.